File: golang/2418_sort_people/sort_people.go
/*
 * @Date: 2023-04-25
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-04-25
 * @FilePath: /algorithm/golang/2418_sort_people/sort_people.go
 */

// Package main ...
package main

import (
	"reflect"
	"sort"
)

func sortPeople(names []string, heights []int) []string {
	n := len(names)
	arr := make([][2]int, n)
	for i, h := range heights {
		arr[i] = [2]int{h, i}
	}
	sort.Slice(arr, func(i, j int) bool { return arr[i][0] > arr[j][0] })
	ans := make([]string, n)
	for i, x := range arr {
		ans[i] = names[x[1]]
	}
	return ans
}

func main() {
	assert := func(a, b []string) {
		if !reflect.DeepEqual(a, b) {
			panic("Not Passed")
		}
	}

	{
		names := []string{"Mary", "John", "Emma"}
		heights := []int{180, 165, 170}
		ans := []string{"Mary", "Emma", "John"}
		assert(sortPeople(names, heights), ans)
	}

	{
		names := []string{"Alice", "Bob", "Bob"}
		heights := []int{155, 185, 150}
		ans := []string{"Bob", "Alice", "Bob"}
		assert(sortPeople(names, heights), ans)
	}
}

File: cpp/2418_sort_people/sort_people.cpp
/*
 * @Date: 2023-04-25
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-04-25
 * @FilePath: /algorithm/cpp/2418_sort_people/sort_people.cpp
 */

#include <cassert>
#include <numeric>
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<string> sortPeople(vector<string>& names, vector<int>& heights) {
    int n = names.size();
    vector<pair<int, int>> arr;
    for (int i = 0; i < n; ++i) {
      arr.emplace_back(-heights[i], i);
    }
    sort(arr.begin(), arr.end());
    vector<string> ans;
    for (int i = 0; i < n; ++i) {
      ans.emplace_back(names[arr[i].second]);
    }
    return ans;
  }
};

int main() {
  {
    vector<string> names{"Mary", "John", "Emma"};
    vector<int> heights{180, 165, 170};
    vector<string> ans{"Mary", "Emma", "John"};
    assert(Solution().sortPeople(names, heights) == ans);
  }

  {
    vector<string> names{"Alice", "Bob", "Bob"};
    vector<int> heights{155, 185, 150};
    vector<string> ans{"Bob", "Alice", "Bob"};
    assert(Solution().sortPeople(names, heights) == ans);
  }
}